man/man3/Snake.3
.TH "Snake" 3 "Thu May 2 2019" "Snoke" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Snake \- 
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <snake\&.h>\fP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "bool \fBinit\fP (\fBPoint\fP begin, short \fBdirection\fP, int length)"
.br
.ti -1c
.RI "void \fBsetScheme\fP ()"
.br
.ti -1c
.RI "void \fBgetCoords\fP (std::list< \fBPoint\fP > *currBody)"
.br
.ti -1c
.RI "\fBPoint\fP \fBgetHeadCoords\fP ()"
.br
.ti -1c
.RI "void \fBsetDirection\fP (int dir=0)"
.br
.ti -1c
.RI "short \fBgetDirection\fP ()"
.br
.ti -1c
.RI "bool \fBmove\fP (\fBLabyrinth\fP labyrinth, \fBBall\fP *ball, \fBPoint\fP *change[2], int changeSize)"
.br
.in -1c
.SS "Private Member Functions"

.in +1c
.ti -1c
.RI "short \fBcheckIntersection\fP (\fBPoint\fP check, \fBLabyrinth\fP laryrinth, \fBBall\fP *ball)"
.br
.ti -1c
.RI "short \fBcheckWisely\fP (\fBPoint\fP coords, \fBPoint\fP bcoords)"
.br
.ti -1c
.RI "void \fBmoveHead\fP (\fBPoint\fP p, \fBPoint\fP *change[2])"
.br
.ti -1c
.RI "void \fBmoveBack\fP (\fBPoint\fP p, \fBPoint\fP *cahnge[2])"
.br
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "short \fBdirection\fP"
.br
.ti -1c
.RI "std::list< \fBPointStyle\fP > \fBstyle\fP"
.br
.ti -1c
.RI "std::deque< \fBPoint\fP > \fBsnakeBody\fP"
.br
.in -1c
.SH "Detailed Description"
.PP 
Describes the snake entity of the game 
.PP
\fBParameters:\fP
.RS 4
\fI{short}\fP direction - current direction where the snake is heading; 
.br
\fI{list<PointStyle>}\fP style - an array, describing how should the snake be colored 
.br
\fI{deque<Point>}\fP snakeBody - a deque type array, containing sposition of snake's body segments 
.RE
.PP

.PP
Definition at line 50 of file snake\&.h\&.
.SH "Member Function Documentation"
.PP 
.SS "short Snake::checkIntersection (\fBPoint\fPcheck, \fBLabyrinth\fPlabyrinth, \fBBall\fP *ball)\fC [private]\fP"
Checking the given point for intersections with \fBBall\fP, borders, obstacles 
.PP
\fBParameters:\fP
.RS 4
\fI{Labyrinth}\fP labyrinth - 2-dimensional array defying current state of every point of the game field (blocked or not) 
.br
\fI{Ball*}\fP ball - a pointer to a \fBBall\fP object(to check intersection with) 
.br
\fI{Point*}\fP change - 2-dimensional array of changes needed to be applied to the labyrinth 
.RE
.PP
\fBReturns:\fP
.RS 4
{short} - type of collision 
.RE
.PP

.PP
Definition at line 191 of file snake\&.cpp\&.
.SS "short Snake::checkWisely (\fBPoint\fPcoords, \fBPoint\fPbcoords)\fC [private]\fP"
If the Point of the labyrinth we are checking is not free then decide the type of collision 
.PP
\fBParameters:\fP
.RS 4
\fI{Point}\fP coords - Point we are checking for type of a collision 
.br
\fI{Point}\fP bcoords - Point, containing the coordinates of the \fBBall\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
{short} - type of the collision 
.RE
.PP

.PP
Definition at line 207 of file snake\&.cpp\&.
.SS "void Snake::getCoords (std::list< \fBPoint\fP > *currBody)"
The method to get the snake whole body coordinates(x ,y) without giving the direct access 
.PP
\fBParameters:\fP
.RS 4
\fI{std::list<Point>}\fP currBody - an array where the current snake body is copied 
.RE
.PP

.PP
Definition at line 276 of file snake\&.cpp\&.
.SS "short Snake::getDirection ()"
The method to get the current direction without giving the direct access 
.PP
\fBReturns:\fP
.RS 4
{short} 
.RE
.PP

.PP
Definition at line 255 of file snake\&.cpp\&.
.SS "\fBPoint\fP Snake::getHeadCoords ()"
The method to get the coordinates(x, y) of the snake's head without giving the direct access 
.PP
\fBReturns:\fP
.RS 4
{Point} 
.RE
.PP

.PP
Definition at line 264 of file snake\&.cpp\&.
.SS "bool Snake::init (\fBPoint\fPbegin, shortdir, intlength)"
Initializes snake 
.PP
\fBParameters:\fP
.RS 4
\fI{Point}\fP begin - starting Point of a snake(where the tail segment will be situated) 
.br
\fI{short}\fP dir - direction of snake's 'growth' as well as it's starting direction 
.br
\fI{int}\fP length - the length of a 'new born' snake 
.RE
.PP
\fBReturns:\fP
.RS 4
{bool} - mark of whether the snake is successfully initialized 
.RE
.PP

.PP
Definition at line 10 of file snake\&.cpp\&.
.SS "bool Snake::move (\fBLabyrinth\fPlabyrinth, \fBBall\fP *ball, \fBPoint\fP *change[2], intchangeSize)"
The snake movement on the game field(should be called in each iteration of the game cycle, unless the snake is 'dead') 
.PP
\fBParameters:\fP
.RS 4
\fI{Labyrinth}\fP labyrinth - the current state of the labyrinth object for intersection checking 
.br
\fI{Ball*}\fP ball - a pointer to a \fBBall\fP object(to check intersection with) 
.br
\fI{Point*}\fP change - 2-dimensional array of changes needed to be applied to the labyrinth 
.RE
.PP
\fBReturns:\fP
.RS 4
{bool} - mark of whether there was a non-boundary non-ball collision 
.RE
.PP

.PP
Definition at line 61 of file snake\&.cpp\&.
.SS "void Snake::moveBack (\fBPoint\fPp, \fBPoint\fP *change[2])\fC [private]\fP"
Checking if we need to remove the back of the snake from the labyrinth(we don't in case it has eaten the \fBBall\fP) 
.PP
\fBParameters:\fP
.RS 4
\fI{Point}\fP p - the desired position of movement 
.br
\fI{Point*}\fP change - 2-dimensional array of changes needed to be applied to the labyrinth 
.RE
.PP

.PP
Definition at line 175 of file snake\&.cpp\&.
.SS "void Snake::moveHead (\fBPoint\fPp, \fBPoint\fP *change[2])\fC [private]\fP"
Moving snake head to a described by parameters position and updating the addition to the labyrinth 
.PP
\fBParameters:\fP
.RS 4
\fI{Point}\fP p - the new head position 
.br
\fI{Point*}\fP change - 2-dimensional array of changes needed to be applied to the labyrinth 
.RE
.PP

.PP
Definition at line 164 of file snake\&.cpp\&.
.SS "void Snake::setDirection (intdirection = \fC0\fP)"
The method to set the direction where the snake is heading 
.PP
\fBParameters:\fP
.RS 4
\fI{int}\fP direction - direction of the snake we are trying to set 
.RE
.PP

.PP
Definition at line 236 of file snake\&.cpp\&.
.SS "void Snake::setScheme ()"

.SH "Member Data Documentation"
.PP 
.SS "short Snake::direction\fC [private]\fP"

.PP
Definition at line 53 of file snake\&.h\&.
.SS "std::deque<\fBPoint\fP> Snake::snakeBody\fC [private]\fP"

.PP
Definition at line 55 of file snake\&.h\&.
.SS "std::list<\fBPointStyle\fP> Snake::style\fC [private]\fP"

.PP
Definition at line 54 of file snake\&.h\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for Snoke from the source code\&.


Libraries/Snake/snake.cpp
/**
 * @file snake.cpp
 * @author Yegor Ignatov
 * @brief realisation of Snake class
 * @version 0.1
 * @date 2019-05-03
 *
 * @copyright Copyright (c) 2019
 *
 */
#include "snake.h"

/**
 * @brief   Initializes snake
 * @param   begin  - starting Point of a snake(where the tail segment will be situated)
 * @param   dir    - direction of snake's 'growth' as well as it's starting direction
 * @param   length - the length of a 'new born' snake
 * @return         - mark of whether the snake is successfully initialized
 */
bool Snake::init(Point begin, short dir, int length)
{
    begin.style.letter = '*';
    if (dir == MVLEFT || dir == MVRIGHT || dir == MVUP || dir == MVDOWN)
    {
        /*
         * Pointer to the variable we want to change, depending on the dir variable
         * @type {short*}
         */
        short* changable;
        if (dir == MVLEFT || dir == MVRIGHT)
        {
            changable = &begin.x;
        }
        else
        {
            changable = &begin.y;
        }

        /*
         * The amount we will change the changable variable, depending on the dir variable
         * @type {short}
         */
        short value;
        if (dir == MVRIGHT || dir == MVDOWN)
        {
            value = 1;
        }
        else
        {
            value = -1;
        }

        for (int i = 0; i < length; i++)
        {
            snakeBody.push_front(begin);
            *changable += value;
        }
        direction = dir;
        return false;
    }
    return true;
}

/**
 * @brief   The snake movement on the game field(should be called in each iteration of the game cycle, unless the snake is 'dead')
 * @param   labyrinth  - the current state of the labyrinth object for intersection checking
 * @param   ball       - a pointer to a Ball object(to check intersection with)
 * @param   change     - 2-dimensional array of changes needed to be applied to the labyrinth
 * @param   changeSize - max(len(change[0]), len(change[1])))
 * @return             - mark of whether there was a non-boundary non-ball collision
 */
bool Snake::move(Labyrinth labyrinth, Ball* ball, Point* change[2], int changeSize)
{
    Point frCoords = snakeBody.front();
    Point bCoords = snakeBody.back();

    /*
     * Getting the future head coords to check intersections
     */
    switch (direction)
    {
        case MVRIGHT:
            frCoords.x++;
            break;
        case MVLEFT:
            frCoords.x--;
            break;
        case MVDOWN:
            frCoords.y++;
            break;
        case MVUP:
            frCoords.y--;
            break;
    }

    /*
     * Flag of whether there was a wall collision and therefor
     * we need to teleport snake's head to the opposite side of the labyrinth
     */
    bool wallFlag = 0;

    /*
     * Checking intersections with borders, obstacles and Ball and deciding where to put the head of the snake
     */
    switch (checkIntersection(frCoords, labyrinth, ball))
    {
        case WALLUP:
            frCoords.y = gameFieldSize.y - 2;
            wallFlag = 1;
            break;
        case WALLBOT:
            frCoords.y = 1;
            wallFlag = 1;
            break;
        case WALLLEFT:
            frCoords.x = gameFieldSize.x - 2;
            wallFlag = 1;
            break;
        case WALLRIGHT:
            frCoords.x = 1;
            wallFlag = 1;
            break;
        case COLL:
            moveBack(bCoords, change);
            return true;
        case NOCOLL:
            break;
        case BALL:
            bCoords.x = -1;
            bCoords.y = -1;
            change[1][0] = ball->getCoords();
            ball->generateBall(labyrinth, change, changeSize);
            change[0][0] = ball->getCoords();
            break;
    }

    /*
     * If there was a wall collision, then we need to check
     * the new head position of the snake for colliding with Ball or obstacles
     */
    if(wallFlag)
    {
        switch (checkIntersection(frCoords, labyrinth, ball))
        {
            case COLL:
                moveBack(bCoords, change);
                return true;
            case NOCOLL:
                break;
            case BALL:
                bCoords.x = -1;
                bCoords.y = -1;
                change[1][0] = ball->getCoords();
                ball->generateBall(labyrinth, change, changeSize);
                change[0][0] = ball->getCoords();
                break;
        }
    }

    /*
     * If there was no obstacle collision then proceed and add Points to the
     * change array, staging them for addition/removal to/from labyrinth
     */
    moveHead(frCoords, change);
    moveBack(bCoords, change);
    return false;

}

/**
 * @brief   Moving snake head to a described by parameters position and updating the addition to the labyrinth
 * @param   p      - the new head position
 * @param   change - 2-dimensional array of changes needed to be applied to the labyrinth
 */
void Snake::moveHead(Point p, Point* change[2])
{
    change[0][1] = p;
    snakeBody.push_front(p);
}

/**
 * @brief   Checking if we need to remove the back of the snake from the labyrinth(we don't in case it has eaten the Ball)
 * @param   p      - the desired position of movement
 * @param   change - 2-dimensional array of changes needed to be applied to the labyrinth
 */
void Snake::moveBack(Point p, Point* change[2])
{
    change[1][1] = p;
    if(p.x != -1 && p.y != -1)
    {
        snakeBody.pop_back();
    }
}

/**
 * @brief   Checking the given point for intersections with Ball, borders, obstacles
 * @param   check     - a Point to check
 * @param   labyrinth - 2-dimensional array defying current state of every point of the game field (blocked or not)
 * @param   ball      - a pointer to a Ball object(to check intersection with)
 * @return            - type of collision
 */
short Snake::checkIntersection(Point check, Labyrinth labyrinth, Ball* ball)
{
    if (!labyrinth.isFree(check))
    {
        return checkWisely(check, ball->getCoords());
    }
    return NOCOLL;

}

/**
 * @brief   If the Point of the labyrinth we are checking is not free then decide the type of collision
 * @param   coords  - Point we are checking for type of a collision
 * @param   bcoords - Point, containing the coordinates of the Ball
 * @return          - type of the collision
 */
short Snake::checkWisely(Point coords, Point bcoords)
{
    if(coords.x == bcoords.x && coords.y == bcoords.y)
    {
        return BALL;
    }
    else if (coords.x == gameFieldSize.x - 1)
    {
        return WALLRIGHT;
    }
    else if (coords.x == 0)
    {
        return WALLLEFT;
    }
    else if (coords.y == gameFieldSize.y - 1)
    {
        return WALLBOT;
    }
    else if (coords.y == 0)
    {
        return WALLUP;
    }
    return COLL;
}

/**
 * @brief   The method to set the direction where the snake is heading
 * @param   direction - direction of the snake we are trying to set
 */
void Snake::setDirection(int direction)
{
    if (direction == MVRIGHT || direction == MVLEFT || direction == MVUP || direction == MVDOWN){
        /*
         * If directions are opposite, do nothing
         */
        if (this->direction + direction == 0)
        {
            return;
        }

        this->direction = direction;
    }
}

/**
 * @brief   The method to get the current direction without giving the direct access
 */
short Snake::getDirection()
{
    return this->direction;
}

/**
 * @brief   The method to get the coordinates(x, y) of the snake's head without giving the direct access
 */
Point Snake::getHeadCoords()
{
    Point headCoords;
    headCoords.x = snakeBody.front().x;
    headCoords.y = snakeBody.front().y;
    return headCoords;
}

/**
 * @brief   The method to get the snake whole body coordinates(x ,y) without giving the direct access
 * @param   currBody - an array where the current snake body is copied
 */
void Snake::getCoords(std::list<Point>* currBody)
{
    currBody->assign(snakeBody.begin(), snakeBody.end());
}
